### PR TITLE
software: Add tetris demo

### PR DESCRIPTION
**Pull request by vmedea**
This adds a game, IceTetris, which is just like normal Tetris except your very important mission is to destroy as many blocks of colorful fruit ice as possible ! :grin:.

The game logic is from [tetris4terminals](https://github.com/emard/tetris4terminals) by @emard et al, I've added a few visual effects, but it's still a small and straightforward game and might be useful as an example.

IMG_20201025_133556

### CODE REVIEW

**Comment by emard** [comment]
HI

I'm currently a bit in doubt should I merge this
because it started as a tetris for terminals
(escape sequences for vt100 and similar
serial or remote login)

your work looks good, it uses tiled graphics for
soft core CPU. I think it's better that you fork and
keep developing on your git because if someone
posts issue on your work, I would have to forward
question to you to solve or answer

**Comment by emard** [comment]
HI

I apologize I thought I should accept this pull request to emard's repository.
I encourage you to fork, use it as base for coding and further development :)

**Comment by dan-rodrigues** [comment]
Looks great, thanks for porting this and adding all the little touches

Only thing I noticed was the title text is slightly cut off, but works fine when built with `.ENABLE_FAST_CPU(1)`. I’ll leave a comment with a quick fix.

File: software/tetris/main.c [hunk 1]
```diff
@@ -0,0 +1,999 @@
...
+
+// Efficiently center fixed-length message.
+#define CENTER(y, color, message) scr_puts(y, TILES_W / 2 - (sizeof(message) - 1) / 2, color, message)
+
+/**
+ * Called when the game starts initially.
+ */
+static void do_beginning(void) {
+    const uint16_t blink = PAL_BLINK << SCROLL_MAP_PAL_SHIFT;
+    scr_clear();
```
Comment (dan-rodrigues) [inline software/tetris/main.c:615]: To fix that issue I mentioned this `scr_clear()` call could move out of here and into `main()` before the loop. It looks like the only place where it transitions to `BEGINNING` state:
```C
scr_clear();
state = BEGINNING;
```

**Comment by vmedea** [comment]
> I'm currently a bit in doubt should I merge this because it started as a tetris for terminals (escape sequences for vt100 and similar serial or remote login)

Sure, I understand! this is why I didn't PR it to your repository. I just wanted to mention you because it's based on your work as well :slightly_smiling_face: 

> I apologize I thought I should accept this pull request to emard's repository.
I encourage you to fork, use it as base for coding and further development :)

Thanks for your comments anyhow. I'm not sure how actively I'm going to work on this. It was helpful for me while figuring out how different VDP features work. But next up I have a different game in mind. Also, this is the point where it is still a basic example. if I do add bells and whistles, I'll develop it further in my own repository with games (a structure like https://github.com/dan-rodrigues/super-miyamoto-sprint).

I still intend to add the sound effect! But I don't have audio hooked up to the ULX3S board at all at the moment.

> Only thing I noticed was the title text is slightly cut off, but works fine when built with .ENABLE_FAST_CPU(1). I’ll leave a comment with a quick fix.

I happen to have that enabled  :blush:. I'll test without and try your suggestion, thanks!

**Comment by vmedea** [comment]
There was a similar problem in `do_gameover`. Those functions aren't even supposed to be called every frame.

After realizing it doesn't need to be a bit field at all, I overhauled the state handling, changing to an enumeration of states and making transitions explicit. This also makes it easier to understand what is happening.

And it works without glitches in the slow CPU mode now!

**Comment by dan-rodrigues** [comment]
:+1: Works great, thanks again

**Comment by vmedea** [comment]
Okay I'm glad it does! Thanks for merging!

**Pull request merged**